snippet_parser/core.py

```diff
@@ -289,94 +193,120 @@ def extract(self, wikitext):
-                for marker in markers_in_snippet:
-                    marker.attrib['class'] = CITATION_NEEDED_MARKER_CLASS
-                    lxml_utils.strip_space_before_element(marker)
-
-                length = len(sr.text_content().strip())
-                self.stats.snippet_lengths[length] += 1
-                if minlen < length < maxlen:
-                    snippet = d(lxml.html.tostring(
-                        sr, encoding = 'utf-8', method = 'html')).strip()
-                    snippets_in_section.add(snippet)
-
-            sectitle = ''
-            if i != 0:
-                # Re-parse the section title because fast_parse is
-                # configured to ignore style tags (see above and 
-                # https://github.com/earwig/mwparserfromhell/issues/40),
-                # but we do want to remove them now with strip_code().
-                sectitle = mwparserfromhell.parse(
-                    str(section.get(0).title).strip()).strip_code()
+            snippets_in_section = self._parse_snippets(
+                html, CITATION_NEEDED_MARKER_CLASS, minlen, maxlen)
+            if snippets_in_section is None: continue
+            sectitle = self._parse_section_title(i, section)
             snippets.append([sectitle, list(snippets_in_section)])
         return snippets
 
+    def extract_from_sentences(self, wikitext, sentences):
+        for sentence in sentences:
+            start = wikitext.find(sentence)
+            if start < 0: continue
+            marked = _SENTENCE_MARKER_MARKUP.format(sent = sentence)
+            wikitext = wikitext[:start] + marked + wikitext[start+len(sentence):]
+
+        wikicode = mwparserfromhell.parse(wikitext)
+        sections = wikicode.get_sections(
+            include_lead = True, include_headings = True, flat = True)
+
+        snippets = []
+        minlen, maxlen = self._cfg.snippet_min_size, self._cfg.snippet_max_size
+        for i, section in enumerate(sections):
+            self._remove_reference_group(section)
+            try:
+                html = self._convert_to_html(section)
+            except:
+                continue
+            snippets_in_section = self._parse_snippets(
+                html, SENTENCE_MARKER_CLASS, minlen, maxlen)
+            if snippets_in_section is None: continue
+            sectitle = self._parse_section_title(i, section)
+            snippets.append([sectitle, list(snippets_in_section)])
+        return snippets
+
+    def _remove_reference_group(self, section):
+        for ref in section.filter_tags(matches = lambda t: t.tag == 'ref'):
+            if ref.has('group'):
+                ref.remove('group')
+
+    def _convert_to_html(self, section):
+        params = dict(
+            text = str(section), **self._cfg.html_parse_parameters)
+        return self._wikipedia.parse(params)['parse']['text']['*']
+
+    def _parse_snippets(self, html, marker_class, minlen, maxlen):
```

Comment:
Do I need to break `_parse_snippets` into more methods?

Reply:
Nice, no I think this looks fine.